index, rpc: Coinstatsindex follow-ups

fjahr opened the pull request:
This is a collection of smaller follow-ups to #19521, addressing several post-merge review comments.

fjahr committed:
index: Avoid unnecessary type casts in coinstatsindex

fjahr committed:
index: Use batch writing in coinstatsindex WriteBlock

fjahr committed:
Index: Return early from failed coinstatsindex init

fjahr committed:
rpc: Add missing gettxoutsetinfo help docs

fjahr committed:
rpc: Block until synced if coinstatsindex is used in gettxoutsetinfo

During initial sync after startup the gettxoutsetinfo RPC will still return an error while catching up. However, after the initial sync the index will not error immediately anymore when it's in the process of syncing to the tip while being called. Instead it will block until synced and then return the response.

fjahr committed:
Index: Improve logging in coinstatsindex

More accurate logging of a warning should make clear if the recovery condition was hit while catching the results of the previous block.

Talkless commented on src/rpc/blockchain.cpp:
```diff
@@ -1117,13 +1117,13 @@ static RPCHelpMan gettxoutsetinfo()
-                            {RPCResult::Type::STR_AMOUNT, "coinbase", ""},
-                            {RPCResult::Type::STR_AMOUNT, "new_outputs_ex_coinbase", ""},
-                            {RPCResult::Type::STR_AMOUNT, "unspendable", ""},
+                            {RPCResult::Type::STR_AMOUNT, "prevout_spent", "Total value of all prevouts spent in this block"},
```
Not sure if "value" is the best wording choice here. I see "amount" used much more in the codebase, and even variables/enums used here in PR uses "amount", not "value".

fjahr replied:
Yeah, I see what you mean and I was on the fence there. I felt like amount could be easier mistaken to mean 'count' in this context but I am not a native speaker so maybe that feeling is just off. I have changed it to 'amount' since, as you say, consistency is better, all other things being equal.

MarcoFalke commented on src/index/coinstatsindex.cpp:
```diff
@@ -24,10 +24,10 @@ struct DBVal {
+    CAmount total_unspendable_amount;
+    CAmount total_prevout_spent_amount;
+    CAmount total_new_outputs_ex_coinbase_amount;
+    CAmount total_coinbase_amount;
```
`unspendables` are also "total", but don't mention it. Might be good to remove the "total_" here, or add it to unspendables?

fjahr replied:
Dropped the `total_` everywhere except for the `m_total_amount`, it seemed right to keep it in this case since it's closed to the corresponding legacy name `nTotalAmount` that is also still in use.

fjahr replied:
Dropped the `total_` everywhere except for the `m_total_amount`, it seemed right to keep it in this case since it's closed to the corresponding legacy name `nTotalAmount` that is also still in use.

MarcoFalke commented on src/index/coinstatsindex.cpp:
```diff
@@ -122,9 +122,12 @@ bool CoinStatsIndex::WriteBlock(const CBlock& block, const CBlockIndex* pindex)
         uint256 expected_block_hash{pindex->pprev->GetBlockHash()};
         if (read_out.first != expected_block_hash) {
+            LogPrintf("%s: WARNING: previous block header belongs to unexpected block %s; expected %s\n",
+                      __func__, read_out.first.ToString(), expected_block_hash.ToString());
```
```suggestion
                      read_out.first.ToString(), expected_block_hash.ToString());
```

This is toggled with `-logsourcelocations`, so can skip here.

fjahr replied:
done

MarcoFalke commented on src/index/coinstatsindex.cpp:
```diff
@@ -391,9 +395,12 @@ bool CoinStatsIndex::ReverseBlock(const CBlock& block, const CBlockIndex* pindex
         uint256 expected_block_hash{pindex->pprev->GetBlockHash()};
         if (read_out.first != expected_block_hash) {
+            LogPrintf("%s: WARNING: previous block header belongs to unexpected block %s; expected %s\n",
+                      __func__, read_out.first.ToString(), expected_block_hash.ToString());
```
same

fjahr replied:
done

Sjors commented on src/index/coinstatsindex.cpp:
```diff
@@ -23,11 +23,11 @@ struct DBVal {
     uint64_t transaction_output_count;
     uint64_t bogo_size;
     CAmount total_amount;
-    CAmount total_subsidy;
```
b778ac8f2b12f771e5b8ddcadc35ee267e57cfb0: I think the `total_` prefix is more clear (and consistent with `total_amount` which you wanted to keep). So maybe use the second approach @MarcoFalke suggested:

> unspendables are also "total", but don't mention it. Might be good to remove the "total_" here, or add it to unspendables?

More importantly, can you add comments explaining how these values are used? Their cumulative nature is a bit counter intuitive, though it makes sense for rollbacks.

fjahr replied:
Done, changed the names to all use `total_` and added comments explaining each of the new values in `coinstats.h` a bit more clearly.

Sjors commented on src/rpc/blockchain.cpp:
```diff
@@ -1117,13 +1117,13 @@ static RPCHelpMan gettxoutsetinfo()
+                            {RPCResult::Type::STR_AMOUNT, "prevout_spent", "Total amount of all prevouts spent in this block"},
+                            {RPCResult::Type::STR_AMOUNT, "coinbase", "Coinbase subsidy amount of this block"},
+                            {RPCResult::Type::STR_AMOUNT, "new_outputs_ex_coinbase", "Total amount of new outputs created by this block"},
+                            {RPCResult::Type::STR_AMOUNT, "unspendable", "Total amount of unspendable outputs in this block"},
```
0f9e936e521d7090c876d0c181ca241c56e387ab: `created in this block`

fjahr replied:
Done

Sjors reviewed (commented):
Mostly happy with f3cf0298180680900eabf1cd2827bd854d4cfeda

Talkless commented:
tACK f3cf0298180680900eabf1cd2827bd854d4cfeda. Built on Debian Sid, ran `-regtest` with `-coinstatsindex`, mined some blocks and checked `gettxoutsetinfo` output and executed `feature_coinstatsindex` functional test.

fjahr committed:
scripted-diff: Fix coinstats data member names

Initially these values were 'per block' in an earlier version but were then changed to total values. The names were not updated to reflect that.

-BEGIN VERIFY SCRIPT-
s() { git grep -l "$1" src | xargs sed -i "s/$1/$2/g"; }

s 'm_block_unspendable_amount'              'm_total_unspendable_amount'
s 'm_block_prevout_spent_amount'            'm_total_prevout_spent_amount'
s 'm_block_new_outputs_ex_coinbase_amount'  'm_total_new_outputs_ex_coinbase_amount'
s 'm_block_coinbase_amount'                 'm_total_coinbase_amount'
s 'block_unspendable_amount'                'total_unspendable_amount'
s 'block_prevout_spent_amount'              'total_prevout_spent_amount'
s 'block_new_outputs_ex_coinbase_amount'    'total_new_outputs_ex_coinbase_amount'
s 'block_coinbase_amount'                   'total_coinbase_amount'

s 'unspendables_genesis_block'              'total_unspendables_genesis_block'
s 'unspendables_bip30'                      'total_unspendables_bip30'
s 'unspendables_scripts'                    'total_unspendables_scripts'
s 'unspendables_unclaimed_rewards'          'total_unspendables_unclaimed_rewards'
s 'm_unspendables_genesis_block'            'm_total_unspendables_genesis_block'
s 'm_unspendables_bip30'                    'm_total_unspendables_bip30'
s 'm_unspendables_scripts'                  'm_total_unspendables_scripts'
s 'm_unspendables_unclaimed_rewards'        'm_total_unspendables_unclaimed_rewards'
-END VERIFY SCRIPT-

fjahr committed:
coinstats: Add comments for new coinstatsindex values

Sjors commented:
ACK d6dbb17

fjahr commented:
@MarcoFalke could you check if you are happy with this? :)

jonatack reviewed (commented):
ACK d6dbb17b1608502850e9e5c6ec29569cfcb23237 tested on signet after rebase to current master

Happy to re-ACK for any of the suggestions below, if you're so inclined.

fjahr commented on src/index/coinstatsindex.cpp:
```diff
@@ -391,9 +395,12 @@ bool CoinStatsIndex::ReverseBlock(const CBlock& block, const CBlockIndex* pindex
-                return error("%s: previous block header belongs to unexpected block %s; expected %s",
-                             __func__, read_out.first.ToString(), expected_block_hash.ToString());
+                return error("%s: previous block header not found; expected %s",
+                             __func__, expected_block_hash.ToString());
```
I think we can really only hit this in cases of data corruption (or a bug somewhere else in the code). So that makes it really hard to test. I have made a note to look into it some more and think about how to add coverage without too much effort.

fjahr commented on src/node/coinstats.h:
```diff
@@ -45,15 +45,25 @@ struct CCoinsStats
+    CAmount total_unspendables_bip30{0};
+    //! Total cumulative amount of outputs sent to unspendable scripts (OP_RETURN for example) up until and including this block
+    CAmount total_unspendables_scripts{0};
+    //! Total cumulative amount coins lost due to unclaimed miner rewards up until and including this block
```
done

fjahr commented on src/rpc/blockchain.cpp:
```diff
@@ -1117,13 +1117,13 @@ static RPCHelpMan gettxoutsetinfo()
                             {RPCResult::Type::OBJ, "unspendables", "Detailed view of the unspendable categories",
                             {
-                                {RPCResult::Type::STR_AMOUNT, "genesis_block", ""},
+                                {RPCResult::Type::STR_AMOUNT, "genesis_block", "The Genesis block subsidy is not spendable"},
```
done

fjahr commented:
Addressed @jonatack 's comments.

Sjors commented:
re-utACK 779e638ca9b2b37c247577d225b93ac762b0602f

jonatack commented:
re-ACK 779e638ca9b2b37c247577d225b93ac762b0602f reviewed diff since last review, changes are doc-only; rebased to current master and verified clean debug build, unit tests, and feature_coinstatsindex functional test

Talkless reviewed (approved):
re-utACK 779e638ca9b2b37c247577d225b93ac762b0602f after cosmetic changes.

laanwj commented on src/index/coinstatsindex.cpp:
```diff
@@ -402,24 +409,24 @@ bool CoinStatsIndex::ReverseBlock(const CBlock& block, const CBlockIndex* pindex
         const auto& tx{block.vtx.at(i)};
 
-        for (size_t j = 0; j < tx->vout.size(); ++j) {
+        for (uint32_t j = 0; j < tx->vout.size(); ++j) {
```
Doesn't `vector<…>::size()` return a `size_t`, making this effectively a cast?
(apparently not necessarily: which `size_type` it returns is C++ library dependent)

laanwj commented:
Code review ACK 779e638ca9b2b37c247577d225b93ac762b0602f
